File: net/src/Sdl.Web.PublicContentApi/IPublicContentApi.cs
using System.Collections.Generic;
using Sdl.Web.PublicContentApi.ContentModel;
using Sdl.Web.PublicContentApi.Utils;

namespace Sdl.Web.PublicContentApi
{
    /// <summary>
    /// Public Content Api
    /// </summary>
    public interface IPublicContentApi
    {
        /// <summary>
        /// Holds global context data passed on to PCA service. Note that context data passed
        /// directly to API methods overwrites these values.
        /// </summary>
        IContextData GlobalContextData { get; set; }

        /// <summary>
        /// Specify type of content to return from API. When set to RAW no conversion will take
        /// place otherwise its treated as model data and will go through conversion to type specified
        /// by DefaultModelType
        /// </summary>
        ContentType DefaultContentType { get; set; }

        /// <summary>
        /// Specify model type to return
        /// </summary>
        DataModelType DefaultModelType { get; set; }

        /// <summary>
        /// Specify how tcdl links get rendered
        /// </summary>
        TcdlLinkRendering TcdlLinkRenderingType { get; set; }

        /// <summary>
        /// Specify how the model-service plugin renders links
        /// </summary>
        ModelServiceLinkRendering ModelSericeLinkRenderingType { get; set; }

        ComponentPresentation GetComponentPresentation(ContentNamespace ns, int publicationId, int componentId,
            int templateId, string customMetaFilter, ContentIncludeMode contentIncludeMode, IContextData contextData);

        ComponentPresentationConnection GetComponentPresentations(ContentNamespace ns, int publicationId,
            InputComponentPresentationFilter filter,
            InputSortParam sort,
            IPagination pagination,
            string customMetaFilter, ContentIncludeMode contentIncludeMode, IContextData contextData);

        Page GetPage(ContentNamespace ns, int publicationId, int pageId, string customMetaFilter,
            ContentIncludeMode contentIncludeMode,
            IContextData contextData);

        Page GetPage(ContentNamespace ns, int publicationId, string url, string customMetaFilter,
            ContentIncludeMode contentIncludeMode,
            IContextData contextData);

        Page GetPage(ContentNamespace ns, int publicationId, CmUri cmUri, string customMetaFilter,
            ContentIncludeMode contentIncludeMode,
            IContextData contextData);

        PageConnection GetPages(ContentNamespace ns, IPagination pagination, string url,
            string customMetaFilter, ContentIncludeMode contentIncludeMode, IContextData contextData);

        BinaryComponent GetBinaryComponent(ContentNamespace ns, int publicationId, int binaryId, string customMetaFilter,
            IContextData contextData);

        BinaryComponent GetBinaryComponent(ContentNamespace ns, int publicationId, string url, string customMetaFilter,
            IContextData contextData);

        BinaryComponent GetBinaryComponent(CmUri cmUri, string customMetaFilter, IContextData contextData);

        ItemConnection ExecuteItemQuery(InputItemFilter filter, InputSortParam sort, IPagination pagination,
            string customMetaFilter, ContentIncludeMode contentIncludeMode, bool includeContainerItems,
            IContextData contextData);

        Publication GetPublication(ContentNamespace ns, int publicationId, string customMetaFilter,
            IContextData contextData);

        PublicationConnection GetPublications(ContentNamespace ns, IPagination pagination, InputPublicationFilter filter,
            string customMetaFilter,
            IContextData contextData);

        string ResolvePageLink(ContentNamespace ns, int publicationId, int pageId, bool renderRelativeLink);

        string ResolveComponentLink(ContentNamespace ns, int publicationId, int componentId, int? sourcePageId,
            int? excludeComponentTemplateId, bool renderRelativeLink);

        string ResolveBinaryLink(ContentNamespace ns, int publicationId, int binaryId, string variantId,
            bool renderRelativeLink);

        string ResolveDynamicComponentLink(ContentNamespace ns, int publicationId, int pageId, int componentId,
            int templateId, bool renderRelativeLink);

        PublicationMapping GetPublicationMapping(ContentNamespace ns, string url);

        dynamic GetPageModelData(ContentNamespace ns, int publicationId, string url,
            PageInclusion pageInclusion, ContentIncludeMode contentIncludeMode,
            IContextData contextData);

        dynamic GetPageModelData(ContentNamespace ns, int publicationId, int pageId,
            PageInclusion pageInclusion, ContentIncludeMode contentIncludeMode, IContextData contextData);

        dynamic GetEntityModelData(ContentNamespace ns, int publicationId, int entityId, int templateId,
            ContentIncludeMode contentIncludeMode, IContextData contextData);

        TaxonomySitemapItem GetSitemap(ContentNamespace ns, int publicationId, int descendantLevels,
            IContextData contextData);

        List<TaxonomySitemapItem> GetSitemapSubtree(ContentNamespace ns, int publicationId, string taxonomyNodeId,
            int descendantLevels, Ancestor ancestor,
            IContextData contextData);
    }
}

File: java/pca-client/src/main/java/com/sdl/web/pca/client/IPublicContentApi.java
package com.sdl.web.pca.client;

import com.sdl.web.pca.client.contentmodel.ContentNamespace;
import com.sdl.web.pca.client.contentmodel.ContentType;
import com.sdl.web.pca.client.contentmodel.IContextData;
import com.sdl.web.pca.client.contentmodel.IPagination;
import com.sdl.web.pca.client.contentmodel.InputItemFilter;
import com.sdl.web.pca.client.contentmodel.Page;
import com.sdl.web.pca.client.contentmodel.enums.DataModelType;
import com.sdl.web.pca.client.contentmodel.enums.PageInclusion;
import com.sdl.web.pca.client.exception.PublicContentApiException;

/**
 * This interface enables java clients to connect to the GraphQL Service
 */
public interface IPublicContentApi {

    /**
     * This method can be used to execute the graphQL PageModelById querie.
     *
     * @param ns            specify the Content namespace
     * @param publicationId specify the publication Id.
     * @param pageId        specify the page Id.
     * @param contentType   specify the content type.
     * @param modelType     specify the model type.
     * @param pageInclusion specify the page inclusion.
     * @param renderContent specify the render content.
     * @param contextData   specify the context data.
     * @return The GraphQL JSON string response with data and errors if any.
     */
    <T> T getPageModelData(ContentNamespace ns, int publicationId, int pageId, ContentType contentType, DataModelType modelType, PageInclusion pageInclusion, boolean renderContent, IContextData contextData, Class<T> clazz) throws PublicContentApiException;

    /**
     * This method can be used to execute the graphQL ItemQuery querie using InputItemFilter & Pagination parameters.
     *
     * @param filter
     * @param pagination
     * @return The GraphQL JSON string response with data and errors if any.
     */
    <T> T executeItemQuery(InputItemFilter filter, IPagination pagination, Class<T> clazz) throws PublicContentApiException;

    /**
     * This method can be used to execute the graphQL Sitemap querie using Page & Class model parameters.
     *
     * @param page
     * @param model
     * @return The GraphQL JSON string response with data and errors if any.
     */
    <T> T executeSiteMap(Page page, Class<T> model) throws PublicContentApiException;

    /**
     * This method can be used to execute the graphQL Sitemap querie using ContentNamespace & publicationId parameters.
     *
     * @param ns
     * @param publicationId
     * @return The GraphQL JSON string response with data and errors if any.
     */
    <T> T getSitemap(ContentNamespace ns, int publicationId, Class<T> clazz) throws PublicContentApiException;

    /**
     * This method can be used to execute the graphQL Sitemap querie using ContentNamespace, publicationId, taxonomyNodeId & includeAncestors parameters.
     *
     * @param ns
     * @param publicationId
     * @param taxonomyNodeId
     * @param includeAncestors
     * @return The GraphQL JSON string response with data and errors if any.
     */
    <T> T getSitemapSubtree(ContentNamespace ns, int publicationId, String taxonomyNodeId, boolean includeAncestors, Class<T> clazz) throws PublicContentApiException;

    /**
     * This method can be used to execute the graphQL EntityModelById querie using ContentNamespace, publicationId & entityId parameters.
     *
     * @param ns
     * @param publicationId
     * @param entityId
     * @return The GraphQL JSON string response with data and errors if any.
     */
    <T> T getEntityModelData(ContentNamespace ns, int publicationId, int entityId, Class<T> clazz) throws PublicContentApiException;
}
